### DISCUSSION TITLE
Copilot and microservices/ multiple services

### DISCUSSION THREAD

**Discussion by Nikola-Milovic** [aws/copilot-cli#2717] [Q&A]
Hello everyone, I created an app, and it seems that it is rather difficult to work with multiple services? Besides not having any examples online, ([besides the emoji app ](https://aws.amazon.com/blogs/containers/developing-an-application-based-on-multiple-microservices-using-the-aws-copilot-and-aws-fargate/)), the pipeline and the buildspec is the same for all services? What if I have different requirements for different services? Like different testing setup, different build setup? Is there a way to build all of them seperately eg, keep them separate and still be able to control them from the copilot?

I am not sure I am missing something, as in the emoji app listed above, we have the project structure of 

SRC
|-copilot
|- serviceA
|- serviceB

Where copilot has a single pipeline with codebuild file and that's it. Both services use the same pipeline and build.

**Comment by iamhopaul123** [comment]
Hello @Nikola-Milovic. You can have the following structure to have two separate pipelines for two services:

```
SRC
|- serviceA
  |- copilot
|- serviceB
  |- copilot
```

Where you run `copilot app init` twice in both `serviceA` dir and `serviceB` dir to create two workspaces respectively using the same app name. Then when you run `pipeline init` the pipeline should only include services within the current workspace. Please let me know if anything is not clear to you!

**Reply by Nikola-Milovic** [reply]
Most interesting! What is then shared between the services? Do I have to create an environment once?

I haven't seen this approach yet but it seems to be what I need. Thanks!

**Reply by iamhopaul123** [reply]
Yeah creating environments doesn't require any local changes so you can do it within either of these two workspaces.

**Reply by vlad-yanum-akveo** [reply]
When I tried to create a pipeline for one of the services, the corresponding files were generated under service root folder (e.g. copilot/buildspec, pipeline.jml)
However, the problem is that actual BUILD does not work:
/src/copilot/buildspec.yml: no such file or directory

So, Build engine is trying to find buildspec at the top root copilot folder (I do not have it, I do not need it)

Please help on how to proceed here?

**Reply by efekarakus** [reply]
Hi @vlad-yanum-akveo !

I'm trying to follow the steps taken:
1. You ran `copilot pipeline init` and it created the `copilot/buildspec.yml` and `copilot/pipeline.yml` files 

Were these files pushed upstream? The Build stage of the pipeline would need to have these files available in the remote for it to build.

**Reply by arko7n** [reply]
Does this approach work with a polyrepo? I have separate repositories for ServiceA and ServiceB. How do I use copilot such that the services belong to the same app but with different repositories and pipelines? 

Running `copilot app init` in serviceA creates the buildspec in `/src/copilot/pipelines/serviceA/buildspec.yml` (not checked in). This doesn't work with CodePipeline which expects the buildspec in ServiceA's repo.
